qubes-rpc/qfile-unpacker.c

```diff
@@ -44,31 +45,61 @@ int main(int argc, char ** argv)
     char *home_dir;
     char *incoming_dir_root;
     char *incoming_dir;
-    int uid, ret;
+    uid_t uid = getuid();
     pid_t pid;
     const char *remote_domain;
     char *procdir_path;
     int procfs_fd;
-    int i;
+    int i, ret;
+    int flags = COPY_ALLOW_SYMLINKS | COPY_ALLOW_DIRECTORIES;
 
     if (argc >= 3) {
         errno = 0;
-        uid = strtol(argv[1], NULL, 10);
-        if (errno)
-            gui_fatal("Invalid user ID argument");
+        char buf[21];
+        int r = snprintf(buf, sizeof buf, "%" PRIu64, (uint64_t)uid);
+        if (r < 1 || r >= (int)sizeof(buf))
+            abort(); // snprintf() failed
+        if (strcmp(buf, argv[1]) != 0)
+            gui_fatal("Called with wrong user ID as argument");
         home_dir = prepare_creds_return_dir(uid);
         incoming_dir = argv[2];
     } else {
-        uid = getuid();
         home_dir = prepare_creds_return_dir(uid);
         remote_domain = getenv("QREXEC_REMOTE_DOMAIN");
         if (!remote_domain) {
             gui_fatal("Cannot get remote domain name");
         }
 
+        const char *service = getenv("QREXEC_SERVICE_FULL_NAME");
```

Comment:
Please don't, use a command line switch instead (the `qfile-unpacker uid dir` invocation must still work, but it can easily be supported as `qfile-unpacker [options] [uid dir]`). It's easy since `/etc/qubes-rpc/qubes.Filecopy` is a script.
See task description at https://github.com/QubesOS/qubes-issues/issues/8332

Reply:
I’ll have two cases:

1. `argv[1]` starts with a digit: uses current command line parsing code and does not support new features.  Possibly emits a deprecation warning.
2. `argv[1]` does _not_ start with a digit: uses `getopt_long(3)` to parse options and supports all features.

Reply:
Do you really need two cases? `getopt_long` should be good in any case, and have uid, dir as non-option arguments left after the call (I don't think you need to care about dir name conflicting with an option, it isn't attacker controlled at all)

Reply:
Ok, with one exception:

> The new calling convention will require `--` before non-option arguments, so there is no ambiguity.

No. It can support `--`, but it should not _require_ it.

Reply:
People know that they need to include `--` before non-option arguments that might begin with `-`, but in this case, a non-option argument that begins with an ASCII digit could _also_ cause problems.  Requiring `--` means that people will _always_ include it, and therefore get correct behavior every time.

Reply:
Still, do not require `--`. If somebody wants to include `--`, they can.